### PR TITLE
fix(swarm): implement `ConnectionHandler::poll_close` for combinators

### PR DESCRIPTION
**Pull request by mxinden**
## Description

Follow-up to https://github.com/libp2p/rust-libp2p/pull/4076.

This is especially relevant since `libp2p-swarm-derive` uses `SelectConnectionHandler`.

## Notes & open questions

@thomaseizinger once this is merged, should we backport it to v0.52?

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm/src/behaviour/toggle.rs [hunk 1]
```diff
@@ -369,4 +369,12 @@ where
             }
         }
     }
+
+    fn poll_close(&mut self, cx: &mut Context<'_>) -> Poll<Option<Self::ToBehaviour>> {
+        if let Some(inner) = self.inner.as_mut() {
```
Comment (thomaseizinger) [inline swarm/src/behaviour/toggle.rs]: I think you can `?` here!

File: swarm/src/handler/select.rs [hunk 2]
```diff
@@ -271,6 +283,24 @@ where
         Poll::Pending
     }
 
+    fn poll_close(&mut self, cx: &mut Context<'_>) -> Poll<Option<Self::ToBehaviour>> {
+        loop {
+            match self.closing_state {
```
Comment (thomaseizinger) [inline swarm/src/handler/select.rs]: Why do we need the closing state? Can't we just proceed to closing `proto2` after the first one returned `Poll::Ready(None)`?

**Review by thomaseizinger — COMMENTED** [review]
Oh nice catch!

Did this land in `v0.52`? I thought we only shipped this together with the removal of the `handler` field from the error?

**Comment by thomaseizinger** [comment]
@mxinden I fixed the compile errors and simplified some of the implementations by relying on the now documented requirement for`fuse`-like behaviour. This greatly simplifies things and is safe to rely on IMO.

**Review by mxinden — COMMENTED** [review]
Requiring fuse like behavior sounds good to me.

Thank you for the additional commits.

**Comment by mxinden** [comment]
Mind approving this pull request @thomaseizinger?

**Comment by mxinden** [comment]
> Did this land in `v0.52`? I thought we only shipped this together with the removal of the `handler` field from the error?

I am probably mistaken. Let's ignore `v0.52` in this context then.

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**